packages/lodestar/src/network/attestationService.ts

```diff
@@ -0,0 +1,400 @@
+  metadata: MetadataController;
+}
+
+/**
+ * Manage random (long lived) subnets and committee (short lived) subnets.
+ */
+export class AttestationService implements IAttestationService {
+  private readonly config: IBeaconConfig;
+  private readonly chain: IBeaconChain;
+  private readonly gossip: Eth2Gossipsub;
+  private readonly metadata: MetadataController;
+  private readonly logger: ILogger;
+
+  // Map of random subnets and the slot until they are needed of current fork
+  private randomSubnets: SubnetMap;
+  // Map of random subnets and the slot until they are needed of next fork
+  // TODO: handle multiple hard forks?
+  private nextForkRandomSubnets: SubnetMap | undefined;
+  // Map of committee subnets and the slot until they are needed
+  private committeeSubnets: SubnetMap;
+  // subset of committeeSubnets with exact slot for aggregator
+  private subscribedCommitteeSubnets: SubnetMap;
+
+  private nextForkSubscriptionTimer: NodeJS.Timeout | undefined;
+
+  /**
+   * A collection of seen validators. These dictate how many random subnets we should be
+   * subscribed to. As these time out, we unsubscribe from the required random subnets and update our ENR.
+   * This is a map of validator index and its last active slot.
+   */
+  private knownValidators = new Map<number, Slot>();
+
+  constructor(modules: IAttestationServiceModules) {
+    this.config = modules.config;
+    this.chain = modules.chain;
+    this.gossip = modules.gossip;
+    this.metadata = modules.metadata;
+    this.logger = modules.logger;
+    const currentFork = this.chain.getForkName();
+    this.randomSubnets = new SubnetMap(currentFork);
+    this.committeeSubnets = new SubnetMap(currentFork);
+    this.subscribedCommitteeSubnets = new SubnetMap(currentFork);
+    const nextFork = this.getNextFork();
+    if (nextFork) {
+      this.nextForkRandomSubnets = new SubnetMap(nextFork.name);
+    }
+  }
+
+  start(): void {
+    this.chain.emitter.on(ChainEvent.clockSlot, this.onSlot);
+    this.scheduleNextForkSubscription();
+  }
+
+  stop(): void {
+    this.chain.emitter.off(ChainEvent.clockSlot, this.onSlot);
+    if (this.nextForkSubscriptionTimer) {
+      clearTimeout(this.nextForkSubscriptionTimer);
+    }
+  }
+
+  /**
+   * Get all active subnets for the hearbeat.
+   */
+  getActiveSubnets(): number[] {
+    const currentSlot = this.chain.clock.currentSlot;
+    const allSubnets = new Set([
+      ...this.randomSubnets.getActive(currentSlot),
+      ...this.committeeSubnets.getActive(currentSlot),
+    ]);
+    return Array.from(allSubnets);
+  }
+
+  getNextForkRandomSubnets(): number[] {
+    if (this.nextForkRandomSubnets) {
+      return this.nextForkRandomSubnets.getActive(this.chain.clock.currentSlot);
+    }
+    return [];
+  }
+
+  /**
+   * Called from /eth/v1/validator/beacon_committee_subscriptions api when validator is an attester.
+   */
+  validatorSubscriptions(subscriptions: phase0.BeaconCommitteeSubscription[] = []): void {
```

Comment:
might call this function `addValidatorSubscriptions`, `addBeaconCommitteeSubscriptions` or similar.